src/validation_error.rs

```diff
@@ -0,0 +1,290 @@
+use serde::{Serialize};
+use valico::common::error::ValicoError;
+
+use crate::validation_state::ValidationState;
+
+type BoxedValicoError = Box<dyn ValicoError>;
+
+#[derive(Serialize, Debug)]
+pub struct ValidationErrorMetadata {
+    pub code: String,
+    pub path: String,
+    pub title: String,
+    pub detail: Option<String>,
+}
+
+impl ValidationErrorMetadata {
+    pub fn from_valico_error(err: &impl ValicoError) -> Self {
+        ValidationErrorMetadata {
+            code: err.get_code().into(),
+            path: err.get_path().into(),
+            title: err.get_title().into(),
+            detail: err.get_detail().map(|s| s.into()),
+        }
+    }
+}
+
+impl From<&BoxedValicoError> for ValidationErrorMetadata {
+    fn from(err: &BoxedValicoError) -> Self {
+        ValidationErrorMetadata {
+            code: err.get_code().into(),
+            path: err.get_path().into(),
+            title: err.get_title().into(),
+            detail: err.get_detail().map(|s| s.into()),
+        }
+    }
+}
+
+#[derive(Serialize, Debug)]
+pub struct ParseErrorLocation {
+    pub index: usize,
+    pub line: usize,
+    pub column: usize,
+}
+
+impl From<serde_yaml::Location> for ParseErrorLocation {
+    fn from(location: serde_yaml::Location) -> Self {
+        ParseErrorLocation {
+            index: location.index(),
+            line: location.line(),
+            column: location.column(),
+        }
+    }
+}
+
+#[derive(Serialize, Debug)]
+pub struct ParseErrorMetadata {
+    pub code: String,
+    pub title: String,
+    pub detail: String,
+    pub location: Option<ParseErrorLocation>,
+}
+
+impl From<serde_yaml::Error> for ParseErrorMetadata {
+    fn from(err: serde_yaml::Error) -> Self {
+        ParseErrorMetadata {
+            code: "parse_error".into(),
+            title: "Parse Error".into(),
+            detail: err.to_string(),
+            location: err.location().map(|location| location.into()),
+        }
+    }
+}
+
+#[derive(Serialize, Debug)]
+#[serde(untagged)]
+pub enum ValidationError {
+    // Schema Validation Errors
+    WrongTypeSchemaError {
+        meta: ValidationErrorMetadata,
```

Comment:
Is this re-factor still planned, for this or a future PR?

Reply:
I would like to. Tbh I was procrastinating because I wasn't sure what the best way would be, since I wanted a solution which would work well for serialising to JS but also makes it easy to work with in Rust in case this is distributed as a crate.

For me, the current front runner solution would be:

'''rs
enum ValidationError {
  SomeError {
    path: String,
    ...fields
  }
}

impl ValidationError {
  pub fn get_path(): String {
    ...
  }
}
'''

Since this allows matching over error types but also just quickly grabbing the value of a given field depending on the use case.

Let me know if you think this is ok or if you have another suggestion for how we couple refactor this :)